gcli/src/meta/executor.rs

```diff
@@ -0,0 +1,318 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! The WASM executor in this module is just for parsing the state types
+//! of gear programs, some of the host functions are missing logics that
+//! is because they are for the on-chain environment data.
+
+use anyhow::{anyhow, Result};
+use wasmi::{
+    AsContextMut, Caller, Engine, Extern, Func, Linker, Memory, MemoryType, Module, Store,
+};
+
+const PAGE_STORAGE_PREFIX: [u8; 32] = *b"gcligcligcligcligcligcligcligcli";
+
+/// HostState for the WASM executor
+#[derive(Default)]
+pub struct HostState {
+    /// Message buffer in host state.
+    pub msg: Vec<u8>,
+}
+
+macro_rules! func {
+    ($store:tt) => {
+        func!($store,)
+    };
+    ($store:tt, $($ty:tt),* ) => {
+        Func::wrap(
+            &mut $store,
+            move |_caller: Caller<'_, HostState>, $(_: $ty),*| { Ok(()) },
+        )
+    };
+    (@result $store:tt, $($ty:tt),* ) => {
+        Func::wrap(
+            &mut $store,
+            move |_caller: Caller<'_, HostState>, $(_: $ty),*| { 0 },
+        )
+    };
+}
+
+/// Call `metadata` method in the WASM code.
+pub fn call_metadata(wasm: &[u8]) -> Result<Vec<u8>> {
+    execute(wasm, "metadata")
+}
+
+/// Executes the WASM code.
+fn execute(wasm: &[u8], method: &str) -> Result<Vec<u8>> {
+    assert!(gear_lazy_pages_interface::try_to_enable_lazy_pages(
+        PAGE_STORAGE_PREFIX
+    ));
+
+    let engine = Engine::default();
+    let module = Module::new(&engine, wasm).unwrap();
+
+    let mut store = Store::new(&engine, HostState::default());
+    let mut linker = <Linker<HostState>>::new();
+
+    // Execution environment
+    //
+    // TODO: refactor this after #3416.
+    {
+        let memory = Memory::new(store.as_context_mut(), MemoryType::new(256, None)).unwrap();
+        linker.define("env", "memory", Extern::Memory(memory))?;
```

Comment:
why dont you want to iter through SysCall enum as its done in prod backend? except calls you really need

Reply:
When the executor loads a WASM module, it validates all of the imports of the module.

For example, if we don't have imports of `block_height` and `block_timestamp`, we can parse metadata for `new-meta-v1` and `new-meta-v2` but not `new-meta-v3`, because it will be failed on the module validation ( missing imports `block_height` and `block_timestamp` )

Reply:
I mean add bindings to linker from iter of all syscalls instead of their manual declaration here

Reply:
oo got it